### PR TITLE
Courant number calculator and addition of SGS viscosity to diffusive state

### PR DESCRIPTION
**Pull request by slifer50**
-A function was made to calculate the courant number: src/Diagnostics/Courant.jl
-SGS viscosity added to diffusive_state in src/Atmos/Model/AtmosModel.jl as it is needed for the diffusive courant number calculation.
-Minimum working example is example/DGmethods/ex_002_dycoms_Courant.jl

### CODE REVIEW

File: src/Diagnostics/Courant.jl [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+Courdy = dt*MPI.Allreduce(sdiff,MPI.MAX,mpicomm)/Dy^2
+Courdz = dt*MPI.Allreduce(sdiff,MPI.MAX,mpicomm)/Dz^2
+
+@info "courant sound + advection x =",Courssx+Couru
+@info "courant sound + advection y =",Courssy+Courv
+@info "courant sound + advection z =",Courssz+Courw
+@info "courant diffusive x =",Courdx
+@info "courant diffusive y =",Courdy
+@info "courant diffusive z =",Courdz
+
```
Comment (smarras79) [inline src/Diagnostics/Courant.jl]: I would remove this and add the following instead:

`cfl_h  = max(Courssx, Courssy)
    cfl_v  = Courssz
    cflu_h = max(Couru, Courv)
    cflu_v = Courw

    cfldiff_h = max(Courdx, Courdy)
    cfldiff_v = Courdz

    @info @sprintf """Courant numbers:                                                                                                                        
           sound_speed      = %.1e                                                                                                                            
           CFL_h            = %.5e                                                                                                                            
           CFL_v            = %.5e                                                                                                                            
           CFLadv_h         = %.5e                                                                                                                            
           CFLadv_v         = %.5e                                                                                                                            
           CFLdiff_h        = %.5e                                                                                                                            
           CFLdiff_v        = %.5e                                                                                                                            
""" thermoQ[50,1,50] cfl_h cfl_v cflu_h cflu_v cfldiff_h cfldiff_v
`

File: src/Diagnostics/Courant.jl [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+Courssx = dt*MPI.Allreduce(ss,MPI.MAX,mpicomm)/Dx
+Courssy = dt*MPI.Allreduce(ss,MPI.MAX,mpicomm)/Dy
+Courssz = dt*MPI.Allreduce(ss,MPI.MAX,mpicomm)/Dz
+Couru = dt*MPI.Allreduce(sau,MPI.MAX,mpicomm)/Dx
+Courv = dt*MPI.Allreduce(sav,MPI.MAX,mpicomm)/Dy
+Courw = dt*MPI.Allreduce(saw,MPI.MAX,mpicomm)/Dz
+Courdx = dt*MPI.Allreduce(sdiff,MPI.MAX,mpicomm)/Dx^2
+Courdy = dt*MPI.Allreduce(sdiff,MPI.MAX,mpicomm)/Dy^2
+Courdz = dt*MPI.Allreduce(sdiff,MPI.MAX,mpicomm)/Dz^2
+
```
Comment (smarras79) [inline src/Diagnostics/Courant.jl]: Add:

`  `
    cfl_h  = max(Courssx, Courssy)
    cfl_v  = Courssz
    cflu_h = max(Couru, Courv)
    cflu_v = Courw
    cfldiff_h = max(Courdx, Courdy)
    cfldiff_v = Courdz`

File: src/Diagnostics/Courant.jl [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                          w̅ = localQ[ijk,4,e] / localQ[ijk,1,e]
+                          e̅_tot = localQ[ijk,5,e] / localQ[ijk,1,e]
+                          q̅_tot = localQ[ijk,6,e] / localQ[ijk,1,e]
+
+                          e_int = e̅_tot - 1//2 * (u̅^2 + v̅^2 + w̅^2) - grav * localvgeo[ijk,grid.x3id,e]
+
+                          ts = PhaseEquil(convert(FT, e_int), localQ[ijk,1,e],q̅_tot)
+                          thermoQ[ijk,1,e]=soundspeed_air(ts)
+                          if ss <= thermoQ[ijk,1,e]
+                          ss = abs(thermoQ[ijk,1,e])
```
Comment (smarras79) [inline src/Diagnostics/Courant.jl]: Are you taking the maximum value of speed of sound within the domain by doing this?

File: src/Diagnostics/Courant.jl (hunk 2 above)
Comment (smarras79) [inline src/Diagnostics/Courant.jl]: Replace with:

    ```
max_sound_speed = MPI.Allreduce(ss,MPI.MAX,mpicomm)
    Courssx = dt*max_sound_speed/Dx
    Courssy = dt*max_sound_speed/Dy
    Courssz = dt*max_sound_speed/Dz
    Couru = dt*MPI.Allreduce(sau,MPI.MAX,mpicomm)/Dx
    Courv = dt*MPI.Allreduce(sav,MPI.MAX,mpicomm)/Dy
    Courw = dt*MPI.Allreduce(saw,MPI.MAX,mpicomm)/Dz
    max_sdiff = MPI.Allreduce(sdiff,MPI.MAX,mpicomm)
    Courdx = dt*max_sdiff/Dx^2
    Courdy = dt*max_sdiff/Dy^2
    Courdz = dt*max_sdiff/Dz^2
```

Comment (smarras79) [inline src/Diagnostics/Courant.jl]: Add: 

`dt_exp = cfl_h * min(Dx, Dy, Dz) / max_sound_speed`
`dt_imp = cfl_v * max(Dx, Dy, Dz) / max_sound_speed`

**Comment by akshaysridhar** [comment]
See #712. This PR is now obsolete.

**Pull request closed without merging**